### PR TITLE
8269223: -Xcheck:jni WARNINGs working with fonts on Linux

### PR DESCRIPTION
**Pull request by mkartashev**
Added an `ExceptionCheck()` followed by `ExceptionDescribe()` and `ExceptionClear()` immediately after the Java calls made from the callback function `ReadTTFontFileFunc()` in `freetypeScaler.c`. 

The exception(s) need to be cleared because we're not returning immediately to Java that would've been able to handle them gracefully. And in order not to loose the exception entirely (even though the return value would also indicate an error condition), print out the exception with `ExceptionDescribe()` to aid in debugging.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269223](https://bugs.openjdk.java.net/browse/JDK-8269223): -Xcheck:jni WARNINGs working with fonts on Linux

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to 13b086860601d2ea6601306973828a2c6ea1f4aa
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4572/head:pull/4572` \
`$ git checkout pull/4572`

Update a local copy of the PR: \
`$ git checkout pull/4572` \
`$ git pull https://git.openjdk.java.net/jdk pull/4572/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4572`

View PR using the GUI difftool: \
`$ git pr show -t 4572`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4572.diff">https://git.openjdk.java.net/jdk/pull/4572.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8269223
+ * @summary Verifies that -Xcheck:jni issues no warnings from freetypeScaler.c
+ * @requires os.family == "linux"
```
Comment (mrserb) [inline test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:28]: Can we run this test on all platforms? Since this bug was not found, means we did not cover this code by the tests, and it will be useful to test it even if the code path will be different on other platforms.

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c [hunk 2]
```diff
@@ -137,6 +143,8 @@ static void invalidateJavaScaler(JNIEnv *env,
                                  FTScalerInfo* scalerInfo) {
     freeNativeResources(env, scalerInfo);
     (*env)->CallVoidMethod(env, scaler, invalidateScalerMID);
+    // NB: Exceptions must not be cleared (and therefore no JNI calls performed) after calling this method
```
Comment (mrserb) [inline src/java.desktop/share/native/libfontmanager/freetypeScaler.c:146]: Please split long lines to 80 chars per line

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c [hunk 3]
```diff
@@ -187,6 +195,8 @@ static unsigned long ReadTTFontFileFunc(FT_Stream stream,
                                           scalerInfo->font2D,
                                           sunFontIDs.ttReadBlockMID,
                                           bBuffer, offset, numBytes);
+            // This is a callback, we are not returning immediately to Java and better report exceptions now
+            CHECK_EXCEPTION(env);
```
Comment (mrserb) [inline src/java.desktop/share/native/libfontmanager/freetypeScaler.c:199]: Probably we should report it only if "debugFonts" was set?

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c [hunk 4]
```diff
@@ -46,6 +46,12 @@
 
 #include "fontscaler.h"
 
+#define CHECK_EXCEPTION(env)            \
+    if ((*(env))->ExceptionCheck(env)) { \
+        (*(env))->ExceptionDescribe(env);\
+        (*(env))->ExceptionClear(env);   \
+    }
```
Comment (prrace) [inline src/java.desktop/share/native/libfontmanager/freetypeScaler.c:53]: https://docs.oracle.com/en/java/javase/11/docs/specs/jni/functions.html#exceptiondescribe

"The pending exception is cleared as a side-effect of calling this function"

So you certainly don't need both of these and I would prefer that Describe only be used if really debugging where we think there's a REAL chance of an exception rather than just to keep JNI happy.

And the upcall that is likely (readBlock) itself will log any IOException (and catch it) in the event of an I/O error so I really think this is unlikely to be useful here.

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c [hunk 5]
```diff
@@ -187,6 +195,8 @@ static unsigned long ReadTTFontFileFunc(FT_Stream stream,
                                           scalerInfo->font2D,
                                           sunFontIDs.ttReadBlockMID,
                                           bBuffer, offset, numBytes);
+            // This is a callback, we are not returning immediately to Java and better report exceptions now
```
Comment (prrace) [inline src/java.desktop/share/native/libfontmanager/freetypeScaler.c:198]: I think the comment is un-needed .. since the only reason to call CHECK_EXCEPTION() is because of this.
Same for the other case.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java (hunk 1 above)
Comment (mkartashev) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java]: Sure; dropped the `@requires` clause.

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c (hunk 2 above)
Comment (mkartashev) [reply src/java.desktop/share/native/libfontmanager/freetypeScaler.c]: Done.

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c (hunk 3 above)
Comment (mkartashev) [reply src/java.desktop/share/native/libfontmanager/freetypeScaler.c]: Done.

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c (hunk 4 above)
Comment (mkartashev) [reply src/java.desktop/share/native/libfontmanager/freetypeScaler.c:53]: Yes, makes sense. I made `CHECK_EXCEPTION()` call either `Clear...` or `Describe...` based on the value of `debugFonts`. 
Please, take a look.

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c (hunk 5 above)
Comment (mkartashev) [reply src/java.desktop/share/native/libfontmanager/freetypeScaler.c]: OK, comments removed.

**Comment by mrserb** [comment]
Looks fine to me, I'll run the tests.

**Comment by mkartashev** [comment]
> Looks fine to me, I'll run the tests.

@mrserb Any news on that front?

**Comment by mrserb** [comment]
> @mrserb Any news on that front?

The new test fails on all platforms:
 - Linux/MacOS: The test should be marked as headful, otherwise it may run on headless systems where "No X11 DISPLAY variable was set"
 - WIndows: The jni warnings are generated in the log.

File: src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp [hunk 6]
```diff
@@ -866,7 +866,8 @@ void D3DRQ_FlushBuffer(void *pParam)
 
     if (!JNU_IsNull(env, pFlush->runnable)) {
         J2dTraceLn(J2D_TRACE_VERBOSE, "  executing runnable");
-        JNU_CallMethodByName(env, NULL, pFlush->runnable, "run", "()V");
+        jboolean ignoreException;
+        JNU_CallMethodByName(env, &ignoreException, pFlush->runnable, "run", "()V");
```
Comment (mrserb) [inline src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp:870]: What is the purpose of this change? the only difference is that in the second case the ExceptionCheck will be called, does it affect something?

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp [hunk 7]
```diff
@@ -6569,11 +6569,11 @@ JNIEXPORT void JNICALL
...
     TRY;
-    jclass inputEventClazz = env->FindClass("java/awt/event/InputEvent");
-    CHECK_NULL(inputEventClazz);
-    jmethodID getButtonDownMasksID = env->GetStaticMethodID(inputEventClazz, "getButtonDownMasks", "()[I");
-    CHECK_NULL(getButtonDownMasksID);
-    jintArray obj = (jintArray)env->CallStaticObjectMethod(inputEventClazz, getButtonDownMasksID);
+    jboolean ignoreException;
+    jintArray obj = (jintArray)JNU_CallStaticMethodByName(env, &ignoreException,
+                                                          "java/awt/event/InputEvent",
+                                                          "getButtonDownMasks", "()[I").l;
```
Comment (mrserb) [inline src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:6575]: obj might be null? Can not we just add CHECK_NULL(obj) here?

File: src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp (hunk 6 above)
Comment (mkartashev) [reply src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp:870]: Yes, the `ExceptionCheck()` call will silence the warnings from `-Xcheck:jni`.

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp (hunk 7 above)
Comment (mkartashev) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:6575]: obj indeed might be null, especially since all kinds of things could go wrong during class/method resolution. Added `CHECK_NULL(obj)` here.

**Comment by mkartashev** [comment]
Ping.

**Comment by prrace** [comment]
I'm going to see if the test passes on the internal systems we use which I think will be mostly variout RH/OL systems for this headless test for Linux - and it also will get run on Windows Server of some kind .. I don't think macOS will hit this code path at all for this test.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+ * @summary Verifies that -Xcheck:jni issues no warnings from freetypeScaler.c
+ * @library /test/lib
+ * @key headful
+ * @run main FreeTypeScalerJNICheck
+ */
+import javax.swing.*;
+import java.awt.*;
+import java.awt.geom.Rectangle2D;
+import java.awt.image.*;
+import java.io.*;
```
Comment (prrace) [inline test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:36]: Can we get rid of all these wild card imports ?

File: src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp (hunk 6 above)
Comment (mrserb) [reply src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp:870]: Does it actually suppress the "Xcheck:jni" or it clears a raised exception? If an exception is still "raised" after this call we should do some additional steps to log/clean it.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8269223
+ * @summary Verifies that -Xcheck:jni issues no warnings from freetypeScaler.c
+ * @library /test/lib
+ * @key headful
```
Comment (prrace) [inline test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:29]: What about this test is headful ?

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    public static void runTest() {
+        String families[] = GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
+        JFrame frame = new JFrame("test");
+        BufferedImage bi = new BufferedImage(1, 1, BufferedImage.TYPE_INT_RGB);
+        Graphics2D g2d = bi.createGraphics();
+
+        for (String ff : families)
+        {
+            Font font = Font.decode(ff);
```
Comment (prrace) [inline test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:59]: Gosh, does anyone still use decode() ? I keep forgetting it exists.
You have all the family names, why not just new Font(ff, Font.PLAIN, 12) ?

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java (hunk 9 above)
Comment (prrace) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:29]: oh you are getting metrics for a JFrame ? That's not going to exercise any new font code so is pointless except to make it so the test has to be headful.

File: src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp (hunk 6 above)
Comment (mkartashev) [reply src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp:870]: Yes, the exception will still be "raised" after this call. Since there are no JNI calls (at least those that are forbidden to make with an exception pending) between here and return back to Java, I believe no additional steps are necessary.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java (hunk 10 above)
Comment (mkartashev) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java]: OK, changed to `new Font(...)`.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java (hunk 8 above)
Comment (mkartashev) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java]: Sure, replaced with single-class imports.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java (hunk 9 above)
Comment (mkartashev) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:29]: But `getFontMetrics()` is the primary "entry point" that generated all the JNI warnings in the first place. And I'm also not sure that we could've gotten all the warnings on Windows without `JFrame`.

Comment (prrace) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java:29]: So what's wrong with
g2d.setFont(font);
FontMetrics metrics = g2d.getFontMetrics(font);

?

No frame needed.

File: src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp (hunk 6 above)
Comment (mrserb) [reply src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp:870]: Then I suggest logging them via some of the trace macros.

**Comment by prrace** [comment]
> passes

Looks good. I tried the test as headless and headful which happens to then exercise a broader set of distros than only one or the other the way our systems are set up.

**Review by prrace — APPROVED** [review]

File: src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp (hunk 6 above)
Comment (mkartashev) [reply src/java.desktop/windows/native/libawt/java2d/d3d/D3DRenderQueue.cpp]: Sure, added some logging.

File: test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java (hunk 9 above)
Comment (mkartashev) [reply test/jdk/java/awt/font/JNICheck/FreeTypeScalerJNICheck.java]: Indeed. Changed the test per your suggestions.

**Comment by mrserb** [comment]
Will the updated test case cover the same code paths where the bugs were found and fixed by this change?

**Comment by mkartashev** [comment]
> Will the updated test case cover the same code paths where the bugs were found and fixed by this change?
I believe so - on Linux. Windows-specific changes are mostly covered (I verified `Java_java_awt_Component_initIDs()` (from `awt_Component.cpp`) and `D3DRQ_FlushBuffer()` (from `D3DRenderQueue.cpp`). Not all the `AwtDesktopProperties::Set...Property()` (from awt_DesktopProperties.cpp ) are touched by the test. I originally applied the changes to all the similar code there without verifying that each piece is invoked by the test.

**Review by mrserb — APPROVED** [review]
